src/pmtiles_adapter.js

```diff
@@ -1,14 +1,276 @@
+
+    // Determine requestPayer: Config takes precedence over URL
+    this.requestPayer = configRequestPayer ?? parsed.requestPayer;
+
+    // Determine region: Config takes precedence over URL
+    this.region = configRegion || parsed.region;
+
+    // Create S3 client
+    this.s3Client = this.createS3Client(
+      parsed.endpoint,
+      this.region,
+      profile,
+      this.verbose,
+    );
+  }
+
+  /**
+   * Parses various S3 URL formats into bucket, key, endpoint, region, and profile.
+   * @param {string} url - The S3 URL to parse.
+   * @returns {object} - An object containing bucket, key, endpoint, region, and profile.
+   * @throws {Error} - Throws an error if the URL format is invalid.
+   */
+  parseS3Url(url) {
+    let region = process.env.AWS_REGION || 'us-east-1';
+    let profile = null;
+    let requestPayer = false;
+
+    // Extract profile parameter from URL
+    const profileMatch = url.match(/[?&]profile=([^&]+)/);
+    if (profileMatch) {
+      profile = profileMatch[1];
+    }
+
+    // Extract region parameter from URL
+    const regionMatch = url.match(/[?&]region=([^&]+)/);
+    if (regionMatch) {
+      region = decodeURIComponent(regionMatch[1]);
+    }
+
+    // Extract requestPayer parameter from URL
+    const requestPayerMatch = url.match(/[?&]requestPayer=(true|1)/i);
+    if (requestPayerMatch) {
+      requestPayer = true;
+    }
+
+    // Remove query parameters for cleaner regex matching
+    let cleanUrl = url
+      .replace(/[?&]profile=[^&]+/, '')
+      .replace(/[?&]region=[^&]+/, '')
+      .replace(/[?&]requestPayer=[^&]+/, '')
+      .replace(/\/+$/, '');
+
+    // Format 1: s3+https://endpoint/bucket:key (Contabo-style)
+    // Example: s3+https://eu2.contabostorage.com/mybucket:terrain/tiles.pmtiles
+    const customMatch = cleanUrl.match(
+      /^s3\+(https?):\/\/([^\/]+)\/([^:]+):(.+)$/,
+    );
+    if (customMatch) {
+      return {
+        endpoint: `${customMatch[1]}://${customMatch[2]}`,
+        bucket: customMatch[3],
+        key: customMatch[4],
+        region,
+        profile,
+        requestPayer,
+      };
+    }
+
+    // Format 2: s3://endpoint/bucket/key (generic S3-compatible)
+    // Example: s3://eu2.contabostorage.com/mybucket/terrain/tiles.pmtiles
+    const endpointMatch = cleanUrl.match(/^s3:\/\/([^\/]+)\/([^\/]+)\/(.+)$/);
+    if (endpointMatch) {
+      return {
+        endpoint: `https://${endpointMatch[1]}`,
+        bucket: endpointMatch[2],
+        key: endpointMatch[3],
+        region,
+        profile,
+      };
+    }
+
+    // Format 3: s3://bucket/key (AWS S3 default)
+    // Example: s3://my-bucket/path/to/tiles.pmtiles
+    const awsMatch = cleanUrl.match(/^s3:\/\/([^\/]+)\/(.+)$/);
+    if (awsMatch) {
+      return {
+        endpoint: null, // Use default AWS endpoint
+        bucket: awsMatch[1],
+        key: awsMatch[2],
+        region,
+        profile,
+      };
+    }
+
+    throw new Error(`Invalid S3 URL format: ${url}`);
+  }
```

Comment:
Would this not be much simpler with [Url](https://nodejs.org/api/url.html)?

The regexing around seems... Suboptimal for maintainability